### PR TITLE
InlineRefactoring: rename inlExpLst to InlineContext

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Rename inlExpLst to InlineContext, and move the code for it into
the inlining files. Expand the context to capture the observation
leading to success. Defer generating the method names until the
context is dumped to avoid unnecessary debug/check allocations.

Make sure observation field on GenTreeCall is properly initialized
and copied, and give it the right type via an opaque forward
declaration.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib one more step along the refactoring road... getting very close to reporting inlining failures in the tree view. PTAL.

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib anyone up for taking a look? Should be relatively quick.

**Comment by pgavlin** [comment]
LGTM

File: src/jit/inline.h [hunk 1]
```diff
@@ -13,6 +13,7 @@
 // InlineTarget -- enum, target of a particular observation
 // InlineImpact -- enum, impact of a particular observation
 // InlineObservation -- enum, facts observed when considering an inline
+// InlineContext -- class, remembers what inlines happened
```
Comment (CarolEidt) [inline src/jit/inline.h]: Would it make sense to group (and comment) these according to which can have multiple instances per candidate, which are one per-inline-candidate and which are per-compile (per-caller)?

Comment (CarolEidt) [inline src/jit/inline.h]: (Not critical - could be part of a future change).

**Comment by CarolEidt** [comment]
LGTM

**Comment by briansull** [comment]
LGTM

File: src/jit/inline.h (hunk 1 above)
Comment (AndyAyersMS) [inline src/jit/inline.h]: Sounds like a good idea. I have more edits coming here so will add this onto the next bit of work.

**Comment by AndyAyersMS** [comment]
Odd, something unexpectedly triggered a new CI run.... guess I'll wait for this though I am sure the previous one was green.

**Pull request merged**